### PR TITLE
Getting Started with an Observable Counter

### PR DESCRIPTION
**Pull request by victlu**
Show example of how to create an Observable Counter.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs [hunk 1]
```diff
@@ -43,14 +47,21 @@ public void Update<T>(T value)
...
-                    if (val < 0)
+                    if (val < 0 && this.isMonotonicValue)
                     {
                         // TODO: log?
                         // Also, this validation can be done in earlier stage.
                     }
                     else
                     {
-                        this.sumLong += val;
+                        if (this.isDeltaValue)
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs:57]: without corresponding changes in the Collect() as well, we wont get cumulative/delta outputs.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs [hunk 2]
```diff
@@ -26,11 +26,15 @@ internal class SumMetricAggregatorLong : IAggregator
...
+        private bool isDeltaValue;
+        private bool isMonotonicValue;
 
-        internal SumMetricAggregatorLong(string name, string description, string unit, Meter meter, DateTimeOffset startTimeExclusive, KeyValuePair<string, object>[] attributes)
+        internal SumMetricAggregatorLong(string name, string description, string unit, Meter meter, DateTimeOffset startTimeExclusive, KeyValuePair<string, object>[] attributes, bool isDeltaValue, bool isMonotonicValue)
         {
             this.startTimeExclusive = startTimeExclusive;
             this.sumMetricLong = new SumMetricLong(name, description, unit, meter, startTimeExclusive, attributes);
+            this.isDeltaValue = isDeltaValue;
+            this.isMonotonicValue = isMonotonicValue;
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs:37]: this is not required to be part of this PR right? (With UpDownCounter being present, we can make this hardcoded for now, i think)

Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs:37]: It sets up the prototype. But, the callers all have it hardcoded to true (MapToMetrics).

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs (hunk 1 above)
Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs:57]: That part was being addressed in 2173. But I guess we're abandoning that and moving it into this PR.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs (hunk 2 above)
Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregatorLong.cs]: I though .NET counter can take non-monotonic values.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            // TODO: Confirm that this approach of
+            // re-using the same instance is correct.
+            // This avoids allocating a new instance.
+            // It is read only for Exporters,
+            // and also there is no parallel
+            // Collect allowed.
+            return isDouble ? (IMetric)this.sumMetricDouble : (IMetric)this.sumMetricLong;
+        }
+
+        public string ToDisplayString()
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:238]: this can be removed

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                }
+            }
+
+            // TODO: Confirm that this approach of
+            // re-using the same instance is correct.
+            // This avoids allocating a new instance.
+            // It is read only for Exporters,
+            // and also there is no parallel
+            // Collect allowed.
+            return isDouble ? (IMetric)this.sumMetricDouble : (IMetric)this.sumMetricLong;
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:235]:
```suggestion
            return isDouble ? this.sumMetricDouble : this.sumMetricLong;
```

File: docs/metrics/getting-started-observable/Program.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+{
+    private static readonly Meter MyMeter = new Meter("TestMeter", "0.0.1");
+
+    public static async Task Main(string[] args)
+    {
+        using var meterProvider = Sdk.CreateMeterProviderBuilder()
+                .AddSource("TestMeter")
+                .AddConsoleExporter((option) =>
+                {
+                    option.IsDelta = false;
```
Comment (cijothomas) [inline docs/metrics/getting-started-observable/Program.cs:36]: lets keep the defaults here, to have as minimum lines as possible in getting started guides.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+                {
+                    sumLong = this.sumLong;
+                    sumDouble = this.sumDouble;
+                }
+
+                isDouble = sumDouble != 0;
+
+                if (isDouble)
+                {
+                    // Include any long values also
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:212]: why include long values?

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                        lock (this.lockUpdate)
+                        {
+                            this.sumLong = val;
+                        }
+                    }
+                }
+            }
+        }
+
+        public IMetric Collect(DateTimeOffset dt, bool collectDelta)
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:149]: in a future PR, we'll remove this CollectDelta option from the Collect, and make it a ctor() parameter. 
And rename 2 similar fields:
CollectDelta - to something to indicate this determines if the Output should be delta or not.
this.isDeltaValue - to something that indicates this indicate if the incoming value is delta or cumulative.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+
+namespace OpenTelemetry.Metrics
+{
+    internal class SumMetricAggregator : IAggregator
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:23]: merging back Double/Long aggregators into one - please make it a separate PR. its not required for adding support for async counters.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        public void Update<T>(T value)
+            where T : struct
+        {
+            long val = 0;
+            double dval = 0;
+            bool isDoubleValue = false;
+
+            // TODO: Confirm the following casts doesn't cause boxing.
+            // VL: It is expected the Compiler will optimize this heavily!
+            if (typeof(T) == typeof(long))
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:60]: only check for Long vs Double. Rest is handled by the provider already.

File: src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+                    // Updating with a Delta Value
+
+                    if (isDoubleValue)
+                    {
+                        if (dval != 0)
+                        {
+                            lock (this.lockUpdate)
+                            {
+                                // TODO: Replace Lock with Interlocked.Add
+                                // VL: We have multiple fields to lock/protect,
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs:109]: will discuss separately - i only see one value to be updated - either a double or a long. so interlocked for long, and try-spin-add for double should work. (will address this separate PR)

**Comment by cijothomas** [comment]
Closing as the core logic from this is copied and merged in https://github.com/open-telemetry/opentelemetry-dotnet/pull/2277
